pkg/kf/push.go

```diff
@@ -120,58 +117,71 @@ func (p *pusher) setupConfig(appName, srcImage string, opts []PushOption) (pushC
 	return cfg, nil
 }
 
-func (p *pusher) imageName(appName string, srcCodeImage bool) string {
-	var prefix string
-	if srcCodeImage {
-		prefix = "src-"
-	}
-	return fmt.Sprintf("%s%s:%d", prefix, appName, time.Now().UnixNano())
+// AppImageName gets the image name for an application.
+func AppImageName(namespace, appName string) string {
+	return fmt.Sprintf("app-%s-%s:%d", namespace, appName, time.Now().UnixNano())
 }
 
-const (
-	buildAPIVersion = "build.knative.dev/v1alpha1"
-)
+// SourceImageName gets the image name for source code for an application.
+func SourceImageName(namespace, appName string) string {
+	return fmt.Sprintf("src-%s-%s:%d", namespace, appName, time.Now().UnixNano())
+}
+
+// JoinRepositoryImage joins a repository and image name.
+func JoinRepositoryImage(repository, imageName string) string {
+	return fmt.Sprintf("%s/%s", repository, imageName)
+}
+
+// BuildName gets a build name based on the current time.
+// Build names are limited by Knative to be 64 characters long.
+func BuildName() string {
+	return fmt.Sprintf("build-%d", time.Now().UnixNano())
+}
 
 func (p *pusher) buildSpec(
+	namespace string,
 	appName string,
 	srcImage string,
 	containerRegistry string,
 	serviceAccount string,
 	buildpack string,
-) (*serving.RawExtension, string, error) {
-	imageName := path.Join(containerRegistry, p.imageName(appName, false))
+) (string, error) {
+	appImageName := AppImageName(namespace, appName)
+	imageDestination := JoinRepositoryImage(containerRegistry, appImageName)
 
 	args := map[string]string{
-		"IMAGE": imageName,
+		"IMAGE": imageDestination,
 	}
 
 	if buildpack != "" {
 		args["BUILDPACK"] = buildpack
 	}
 
-	// Knative Build wants a Build, but the RawExtension (used by the
-	// Configuration object) wants a BuildSpec. Therefore, we have to manually
-	// create the required JSON.
-	buildSpec := builds.PopulateTemplate(
-		"", // no name provided
-		build.TemplateInstantiationSpec{Name: "buildpack", Kind: build.ClusterBuildTemplateKind},
+	buildName := BuildName()
+
+	if _, err := p.buildClient.Create(
+		buildName,
+		builds.BuildpackTemplate(),
 		builds.WithCreateServiceAccount(serviceAccount),
 		builds.WithCreateArgs(args),
 		builds.WithCreateSourceImage(srcImage),
-		builds.WithCreateNamespace(""), // set blank namespace so Knative can choose
-	)
+		builds.WithCreateNamespace(namespace),
+	); err != nil {
+		return "", err
+	}
 
-	buildSpecRaw, err := json.Marshal(buildSpec)
-	if err != nil {
-		return nil, "", err
+	if err := p.buildClient.Tail(buildName, builds.WithTailNamespace(namespace)); err != nil {
```

Comment:
Have you tried this with a bogus service account?

```
kf push someapp --container-registry gcr.io/something-right --service-account something-wrong
```

There were problems with blocking on build logs in the past.

Reply:
I just validated, the bad service account doesn't seem to block.